sdk/key_vault/src/client.rs

```diff
@@ -1,73 +1,67 @@
 /// use azure_identity::token_credentials::DefaultCredential;
 /// let creds = DefaultCredential::default();
-/// let client = KeyVaultClient::new(&creds, &"test-key-vault");
+/// let client = KeyClient::with_name(&"test-key-vault", &creds).unwrap();
 /// ```
 #[derive(Debug)]
-pub struct KeyVaultClient<'a, T> {
+pub struct KeyClient<'a, T> {
+    pub(crate) vault_url: Url,
+    pub(crate) endpoint: String,
     pub(crate) token_credential: &'a T,
-    pub(crate) keyvault_name: &'a str,
-    pub(crate) endpoint_suffix: String,
-    pub(crate) keyvault_endpoint: String,
     pub(crate) token: Option<TokenResponse>,
 }
 
-impl<'a, T: TokenCredential> KeyVaultClient<'a, T> {
-    /// Creates a new `KeyVaultClient` with an endpoint suffix. Useful for non-public Azure clouds.
-    /// For the default public environment, use `KeyVaultClient::new`.
+impl<'a, T: TokenCredential> KeyClient<'a, T> {
+    /// Creates a new `KeyClient`.
     ///
     /// # Example
     ///
     /// ```no_run
-    /// use azure_key_vault::KeyVaultClient;
+    /// use azure_key_vault::KeyClient;
     /// use azure_identity::token_credentials::DefaultCredential;
     /// let creds = DefaultCredential::default();
-    /// let client = KeyVaultClient::with_endpoint_suffix(&creds, &"test-keyvault", "vault.azure.net".to_owned());
+    /// let client = KeyClient::new("test-key-vault.vault.azure.net", &creds).unwrap();
     /// ```
-    pub fn with_endpoint_suffix(
-        token_credential: &'a T,
-        keyvault_name: &'a str,
-        endpoint_suffix: String,
-    ) -> Self {
-        let endpoint = format!("https://{}.{}", keyvault_name, endpoint_suffix);
-        Self {
+    pub fn new(vault_url: &str, token_credential: &'a T) -> Result<Self> {
+        let vault_url = Url::parse(vault_url)?;
+        let endpoint = endpoint_suffix(&vault_url)?;
+        let client = KeyClient {
+            vault_url,
+            endpoint,
             token_credential,
-            keyvault_name,
-            endpoint_suffix,
-            keyvault_endpoint: endpoint,
             token: None,
-        }
+        };
+        Ok(client)
     }
 
-    /// Creates a new `KeyVaultClient`.
+    /// Creates a new `KeyClient` from provided vault name
     ///
     /// # Example
     ///
     /// ```no_run
-    /// use azure_key_vault::KeyVaultClient;
+    /// use azure_key_vault::KeyClient;
     /// use azure_identity::token_credentials::DefaultCredential;
     /// let creds = DefaultCredential::default();
-    /// let client = KeyVaultClient::new(&creds, &"test-keyvault");
+    /// let client = KeyClient::with_name("test-key-vault", &creds).unwrap();
     /// ```
-    pub fn new(token_credential: &'a T, keyvault_name: &'a str) -> Self {
-        KeyVaultClient::with_endpoint_suffix(
-            token_credential,
-            keyvault_name,
-            PUBLIC_ENDPOINT_SUFFIX.to_owned(),
-        )
+    pub fn with_name(vault_name: &str, token_credential: &'a T) -> Result<Self> {
```

Comment:
This will only work in the public cloud - not other environments. I don't think it's worth included at all. I'll open a separate issue to remove it. In general, we want to avoid anything that fixes a client to a single environment. Quite a few (including KV) have different domain suffixes in different clouds.

Reply:
https://github.com/Azure/azure-sdk-for-rust/issues/201